Honor special locations in terrain mask

GregoryLundberg opened the pull request:
Added overlay_special_locations=yes|no, default no, to [terrain_mask]. When yes, side starting locations and special locations marked in the mask replace those in the effected area of the underlying map by removing any from the underlying map, and applying only those given in the mask.

gfgtdf commented:
If we the special locations ignore the [rule] stuf anyways shoudl maybe move them out of the xy loop aswell, to me the code currently looks like

```
for each location do {
  ...
  if (overlay_special_locations) {
    foreach spaecial location in other map in loc do {
      add_special_location();
    }
  }
}
```
i think it woudl be faster if we did

```
for each location do {
  ...
}
if (overlay_special_locations) {
  foreach spaecial location in other map do {
    add_special_location();
  }
}
```
since we just have to iterate over m.starting_positions_ instead of doing a loookup in it for each location.

gfgtdf commented:
also since overlay_special_locations=yes seems to have no effect unless the mask data actuall any starting locations defined i think its save have it =yes as default.

CelticMinstrel commented on src/map/map.cpp:
```diff
@@ -302,34 +303,38 @@ void gamemap::overlay(const gamemap& m, const config& rules_cfg, int xpos, int y
+			if (overlay_special_locations) {
+				starting_positions_.right.erase(t_translation::coordinate(x2, y2));
+				for (auto& pair : m.starting_positions_.right.equal_range(t_translation::coordinate(x1, y1))) {
+					starting_positions_.insert(tstarting_positions::value_type(pair.second, t_translation::coordinate(x2, y2)));
```
`insert()` fails if the key already exists in the map (as I'm sure you're aware). `starting_positions_[key] = value` would overwrite the location if it already exists. I'm not sure I quite understand your reasoning for not doing this?

GregoryLundberg replied:
Me? I just activated the code which was there but disabled by 'if (false)'

I did some testing and it appears to work. But I will take your comment as a request for testing the case where a location (starting number or special name) exists elsewhere on the map.

gfgtdf replied:
Note that starting_positions_ is a boost::bimaps::bimap so it might not be as easy as starting_positions_[key].

I agree that overwriting would be better here.

GregoryLundberg commented:
I will test re-factoring the location section outside the loop. Frankly, we're only talking about a few hundred, perhaps a thousand or so, tiles; so I doubt we'd be able to measure the performance difference. But, it would be better style and probably a bit more clear. I just need to ensure that all (x,y) locations on the effected map are cleared of all location data before copying the new data from the mask to the map.

I don't know what 1.12 did, if anything. The code I found had 'overlay_special_locations' as the constant 'false' and applied only in some cases where it should apply everywhere. I will take a look at LoW S01, as my test case, under 1.12 but I'll need to refactor a leader from [side] to prestart to even test it .. which makes me think it'll become a moot point ...

gfgtdf commented:
>  I just need to ensure that all (x,y) locations on the effected map are cleared of all location data before copying the new data from the mask to the map.


hmm why woudl you want to do this?

GregoryLundberg commented:
That's the effect of the current code, as I read it. It's done cell by cell.

    starting_positions_.right.erase(t_translation::coordinate(x2, y2));

GregoryLundberg commented:
Comments had me check how 1.12.6 handles this. It's sort of confusing but it boils down to "1.12.6 forgot to add an offset, so it does it, but not how one would expect."  Checking did, however, point out that I need to make some more corrections for this PR.

I will try to explain how it works, and how I think it should work, here, so we can talk it out as I work on it and, hopefully, come up with a brief description which I can put in the changelog or pass to Zookeeper (or whoever) for discussion on the Wiki or forums, as needed.

My main intention, of course, it to clarify my own thoughts, as well as to allow others to suggest changes or improvements as I work on this issue. So bear with me.

Let's start with a map and a mask. I'll use an 8x8 map and a 2x2 mask, offsetting the mask toward the lower-right.

       01234567
      +--------+
    0 |        |
    1 |        |
    2 |  X     |
    3 |        |
    4 |    MM  |
    5 |    ML  |
    6 |      Y |
    7 |        |
      +--------+

M is a mask position which (optionally) changes the tile graphics when viewed in the game; for example Gs. L also changes the graphics, but includes a (leader) position mark. For instance L might read as "1 Gs" saying the leader for side 1 should appear here, and change it to plain green Grass. X and Y mark locations I'll explain in a moment.

The WML for this mask might read something like:

        [terrain_mask]
            x,y=4,4
            mask="border_size=0
    usage=map
    
    Gs, Gs
    Gs, 1 Gs
    "
        [/terrain_mask]

Now, one would expect that, if leader 1 does not specify a position or placement in its [unit], it would appear at the location marked "L" on the map.

But, for 1.12.6, that is not the case! Instead, 1.12.6 places the leader at the "X". Why there? Because 1.12.6 failed to consider the [terrain_mask] offset (x,y) when calculating which tile should receive the leader. So, since the mark was at (1,1) relative to the mask, and 1.12.6 always uses an offset of (1,1), the leader will always appear at "X" (2,2) .. REGARDLESS OF THE OFFSET SPECIFIED.

The idea of the new option "overlay_special_locations" is that the positions, whether present, or not, given in the [terrain_mask] should replace those in the original map. For example, if the map, originally had the leader for side 1 placed at one of the "M" locations, that mark would be removed and the mark would appear at "L".

If we ignore the error in 1.12.6 and do that, we still have an issue: what if the map had originally placed the leader at the "X" mark? The intention of the terrain mask is that it only effects the area it covers. But, if so, we end up with a map showing TWO locations for the leader for side 1!

If we simply create a map with two locations, not using [terrain_mask] at all, we will get a warning on the console (stderr) to that effect, and the "last" (right-most, bottom-most) is the one used.

So, for the example so far, once we apply the mask, if the original map has a leader mark at "X" and the mask specifies one at "L", the game will use the one at "L". And we only get a warning on the console if the mask, itself, specifies a duplicated position.

But what if the map originally specified a leader position below, or to the right of (if on the same line), the leader position given in the mask? For example, the at "Y" mark? While I could not test this on 1.12.6 the expectation is the terrain mask would be ignored and the leader would appear at "Y". Again, no warning would appear on the console since the mask, itself, did not have a duplicate.

So, my intention is to update this PR as follows:

1) Correct the positions for leaders and specials given in [terrain_mask] to honor the offset. For existing (ie., 1.12.6) content which uses a full-map-size terrain mask, this will have no effect. For those which used a sub-set mask, as shown in the example above, this is a BREAKING CHANGE since the leader position will move. But, since, in that case, the leader position was not moved with the mask, it seems unlikely any existing content would use a sub-set sized [terrain_mask] and expect leader placement to be correct. So, while a breaking change, it seems unlikely any content would actually break.

2) Rename "overlay_special_positions" to "enable_special_positions". In the 1.13.5+dev which I am working from, the entire processing of leader and special positions was disabled. This option will re-enable it. If omitted or specified as no, the [terrain_mask] will have no effect upon leader or special positions whatsoever. If specified, this option will add positions but not remove any.

3) Add a second option "clear_existing_positions" which, when set, will remove any pre-existing (from the underlying map, possibly modified by earlier terrain masks) leader and special positions WITH THE SAME NAME. This, of course, has no effect if enable_special_positions is not specified as yes.

So, we have three use cases:

1) "enable_special_positions = no" .. any special positions in the terrain mask are ignored. The mask only changes the terrain.

2) "enable_special_positions = yes" and "clear_existing_positions = no" .. special positions from the terrain mask are ADDED, but none are removed. This can result in duplicated positions; which will be specifically checked and flagged with a console warning. When it occurs, the last position given (usually the one from the mask) will be used and all other duplicates removed.

3) "enable_special_positions = yes" and "clear_existing_positions = yes" .. as above, but no warning can appear since each position is removed from the underlying map prior to being added. Note that the duplicate warning can still occur when the terrain mask, itself, as duplicates. Also not this has no effect upon positions already on the map which are not named in the terrain mask.

gfgtdf commented:
I feel like most of these options are not really needed, becasue
1) When you dotn want to mess with stating locatiosn in your overlays then you usually simply woudln pput them in the map
2) from how i understood it, clear_existing_positions = no would have no effect other than gibing an additional warning.

So what i suggest, is to have just one paramter, that only specifies what happens in the case that one starting positions is defined on both the orignal map and the overlays, not guarneteed this example code works:
```
	for location loop{
		...
	}
	....
	bool overwrite = rules_cfg["overwrite_starting_positions"].to_bool(true);
	for (auto& pair : m.starting_positions_) {
		auto new_pos = pair.right + offset;
		if (overwrite) {
			starting_positions_.left[pair.left] = new_pos;
		}
		else {
			starting_positions_.insert(tstarting_positions::value_type(pair.left, new_pos));
		}
	}

```

GregoryLundberg commented:
I was thinking it terms of how to get from the two starting points we have now to where I think we ought to end up.

1) In 1.12.6 the only usable option is when the mask is the same size as the map. In this case, all map positions are completely erased. The only positions seen are those on the mask. Note, however, that to see this effect, you must remove the leader from the [side] using no_leader=yes and place the leader during an event such as prestart.

2) In 1.13.5+dev, at present, all positions from the mask are ignored in all cases and only those positions on the map take effect.

So I was looking for a way to maintain both implementations .. 1.12.6 (only the mask matters, if you do it right) and 1.13.5 (only the map matters) .. and still offer what seems to be a more rational choice.

Frankly, my first impulse was to simply ignore all history and do it "right" without options .. that is:

    for each position from the mask {
        if the position name exists on the map then {
            delete the name from the map
        }
        add the position name and location to the map
    }

Note that a given tile may have multiple position names but a given name must be unique on the resulting map.

This also maintains the last-vote-wins philosophy we see on a map-by-map basis (and "last" remains undefined).

And not have any flags or options at all. If an existing [terrain_mask] map has positions which have never have taken effect and would be incorrect (ie., LoW S01), and the scenario using it actually makes use of them (LoW does not) then a change to the mask and/or scenario is required.

And, unless [terrain_mask] is elevated to toplevel, the use remains dependent upon the author removing all effected [unit] from [side] and moving them to an event such as prestart.

gfgtdf commented:
well the 1.13.5 'implementation' was rather a WIP so i wouldn't care for comability here. In fact i was the one who changed that code in 1.13 and i'm quite sure that i just simply forgot to implement the starting locations code in overlay and left the if(false) in there.

And what yu described about the wrong offset in 1.12 sounds liek a bug for me, in particular i think we can assume that in 1.12 are no campaigsn that use masks with starting positions with a nontrivial offset.

So i'd just do what your 'first impule' was.

GregoryLundberg commented:
OK. "Short, simple and clean" it is. No new options and not worrying about breaking changes.

GregoryLundberg commented:
Code complete.

Tested using a full-sized [terrain_mask].
Tested using a small [terrain_mask] offset into the result map from above.
Tested using the same small [terrain_mask] using a different offset into the result map from above.

Tested the new tag retrieves both side numbers and string names from the map and the location variable is not set if the named location does not exist.

gfgtdf commented on src/map/map.cpp:
```diff
@@ -325,15 +325,14 @@ void gamemap::overlay(const gamemap& m, const config& rules_cfg, int xpos, int y
+
+	for(auto& pair : m.starting_positions_.left) {
+		starting_positions_.left.erase(pair.first);
+		starting_positions_.insert(tstarting_positions::value_type(pair.first, t_translation::coordinate(pair.second.x + xpos, pair.second.y + ypos)));
```
I think its not as eay as just adding a fixed offset to pair.second, due to the hex-based map, the code in the above loop above also does a more complicated check in the y coordinate to make sure it also wokrs with odd x positions.

Also, you might want to wrap the loob body ina  if(on_map(target loc)) check.

CelticMinstrel commented:
What? No new options? I do think it could be nice to include an option to ignore the mask's special locations.

GregoryLundberg commented:
The case could be made that all you need do is fix the mask in the editor.

I'll put the option back in. This time, though, it'll be a 'disable' default 'no' style ...

GregoryLundberg commented:
Fixed offsets for even/odd columns.
Added ignore_special_locations=yes|no, default no
Tested

Offsetting the mask off-map generates no errors or warnings.
Creating a unit for an off-map location generates no error or warnings.
So I see no reason to warn if the offset moves a location off-map.
It is the same as specifying x,y off-map; which is silent

I believe I had finished with the corrections and comments and this is ready if there are no more issues.
